module/Company/src/Company/Service/Company.php

```diff
@@ -2,52 +2,454 @@
+
+            return $packet;
+        } else {
+            $translator = $this->getTranslator();
+            throw new \User\Permissions\NotAllowedException(
+                $translator->translate('You are not allowed to edit packets')
+            );
+        }
+    }
+
+    /**
+     * Returns all companies with a given $slugName and makes them persistent
+     *
+     * @param mixed $slugName
+     */
+    public function getEditableCompaniesWithSlugName($slugName)
+    {
+        if ($this->isAllowed('edit')) {
+            return $this->getCompanyMapper()->findEditableCompaniesWithSlugName($slugName, true);
+        } else {
+            $translator = $this->getTranslator();
+            throw new \User\Permissions\NotAllowedException(
+                $translator->translate('You are not allowed to edit companies')
+            );
+        }
+    }
+
+    /**
+     * Returns all jobs with a given slugname, owned by a company with 
+     * $companySlugName
+     *
+     * @param mixed $companySlugName
+     * @param mixed $jobSlugName
+     */
+    public function getEditableJobsWithSlugName($companySlugName, $jobSlugName)
+    {
+        if ($this->isAllowed('edit')) {
+            return $this->getJobMapper()->findJobWithSlugName($companySlugName, $jobSlugName);
+        } else {
+            $translator = $this->getTranslator();
+            throw new \User\Permissions\NotAllowedException(
+                $translator->translate('You are not allowed to edit jobs')
+            );
+        }
+    }
+
+    /**
+     * Returns all jobs with a $jobSlugName, owned by a company with a 
+     * $companySlugName
+     *
+     * @param mixed $companySlugName
+     * @param mixed $jobSlugName
+     */
+    public function getJobsWithSlugName($companySlugName, $jobSlugName)
+    {
+        return $this->getJobMapper()->findJobWithSlugName($companySlugName, $jobSlugName);
     }
 
+    /**
+     * Returns all jobs in the database
+     *
+     */
     public function getJobList()
     {
         return $this->getJobMapper()->findAll();
     }
 
+    /**
+     * Get the Company Edit form.
+     *
+     * @return Company Edit form
+     */
+    public function getCompanyForm()
+    {
+        return $this->sm->get('company_admin_edit_company_form');
+    }
+
+    /**
+     * Returns a the form for entering packets
+     *
+     */
+    public function getPacketForm()
+    {
+        return $this->sm->get('company_admin_edit_packet_form');
+    }
+
+    /**
+     * Returns the form for entering jobs
+     *
+     */
+    public function getJobForm()
+    {
+        return $this->sm->get('company_admin_edit_job_form');
+    }
+
+    /**
+     * Returns all jobs that are active
+     *
+     */
     public function getActiveJobList()
     {
         $jl = $this->getJobList();
         $r = array();
-        foreach($jl as $j) {
+        foreach ($jl as $j) {
             if ($j->getActive()) {
                 array_push($r, $j);
```

Comment:
No it doesn't, because it doesn't (clearly) describe it's exact behaviour. When seeing it, you may ask yourself if it pushes elements to the front, or to the back. And you may even wonder which of the arguments is the element on which elements are pushed.

And, in every other place in our code, we use the standard PHP append syntax (which would be `$r[] = $j`).